File: HMCL/src/main/java/org/jackhuang/hmcl/ui/InstallerController.java
/*
 * Hello Minecraft! Launcher.
 * Copyright (C) 2018  huangyuhui <huanghongxun2008@126.com>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see {http://www.gnu.org/licenses/}.
 */
package org.jackhuang.hmcl.ui;

import javafx.fxml.FXML;
import javafx.scene.control.ScrollPane;
import javafx.scene.layout.VBox;
import org.jackhuang.hmcl.download.MaintainTask;
import org.jackhuang.hmcl.download.game.VersionJsonSaveTask;
import org.jackhuang.hmcl.game.GameVersion;
import org.jackhuang.hmcl.game.Library;
import org.jackhuang.hmcl.game.Version;
import org.jackhuang.hmcl.setting.Profile;
import org.jackhuang.hmcl.task.Schedulers;
import org.jackhuang.hmcl.task.Task;
import org.jackhuang.hmcl.ui.download.InstallerWizardProvider;

import java.util.LinkedList;
import java.util.Optional;
import java.util.function.Consumer;

public class InstallerController {
    private Profile profile;
    private String versionId;
    private Version version;
    @FXML
    private ScrollPane scrollPane;
    @FXML private VBox contentPane;
    private String forge;
    private String liteLoader;
    private String optiFine;

    @FXML
    private void initialize() {
        FXUtils.smoothScrolling(scrollPane);
    }

    public void loadVersion(Profile profile, String versionId) {
        this.profile = profile;
        this.versionId = versionId;
        this.version = profile.getRepository().getResolvedVersion(versionId);

        contentPane.getChildren().clear();
        forge = liteLoader = optiFine = null;

        for (Library library : version.getLibraries()) {
            Consumer<InstallerItem> removeAction = x -> {
                LinkedList<Library> newList = new LinkedList<>(version.getLibraries());
                newList.remove(library);
                new MaintainTask(version.setLibraries(newList))
                        .then(variables -> new VersionJsonSaveTask(profile.getRepository(), variables.get(MaintainTask.ID)))
                        .with(profile.getRepository().refreshVersionsAsync())
                        .with(Task.of(Schedulers.javafx(), () -> loadVersion(this.profile, this.versionId)))
                        .start();
            };

            if (library.getGroupId().equalsIgnoreCase("net.minecraftforge") && library.getArtifactId().equalsIgnoreCase("forge")) {
                contentPane.getChildren().add(new InstallerItem("Forge", library.getVersion(), removeAction));
                forge = library.getVersion();
            }
            if (library.getGroupId().equalsIgnoreCase("com.mumfrey") && library.getArtifactId().equalsIgnoreCase("liteloader")) {
                contentPane.getChildren().add(new InstallerItem("LiteLoader", library.getVersion(), removeAction));
                liteLoader = library.getVersion();
            }
            if (library.getGroupId().equalsIgnoreCase("net.optifine") && library.getArtifactId().equalsIgnoreCase("optifine")) {
                contentPane.getChildren().add(new InstallerItem("OptiFine", library.getVersion(), removeAction));
                optiFine = library.getVersion();
            }
        }
    }

    @FXML
    private void onAdd() {
        Optional<String> gameVersion = GameVersion.minecraftVersion(profile.getRepository().getVersionJar(version));

        if (!gameVersion.isPresent())
            Controllers.dialog("version.cannot_read");
        else
            Controllers.getDecorator().startWizard(new InstallerWizardProvider(profile, gameVersion.get(), version, forge, liteLoader, optiFine));
    }
}

File: HMCL/src/main/kotlin/org/jackhuang/hmcl/ui/InstallerController.kt
/*
 * Hello Minecraft! Launcher.
 * Copyright (C) 2017  huangyuhui <huanghongxun2008@126.com>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see {http://www.gnu.org/licenses/}.
 */
package org.jackhuang.hmcl.ui

import javafx.fxml.FXML
import javafx.scene.control.ScrollPane
import javafx.scene.layout.VBox
import org.jackhuang.hmcl.download.game.VersionJSONSaveTask
import org.jackhuang.hmcl.game.Version
import org.jackhuang.hmcl.game.minecraftVersion
import org.jackhuang.hmcl.setting.Profile
import org.jackhuang.hmcl.task.Scheduler
import org.jackhuang.hmcl.task.task
import org.jackhuang.hmcl.ui.download.InstallWizardProvider
import java.util.*

class InstallerController {
    private lateinit var profile: Profile
    private lateinit var versionId: String
    private lateinit var version: Version

    @FXML lateinit var scrollPane: ScrollPane
    @FXML lateinit var contentPane: VBox

    private var forge: String? = null
    private var liteloader: String? = null
    private var optifine: String? = null

    fun initialize() {
        scrollPane.smoothScrolling()
    }

    fun loadVersion(profile: Profile, versionId: String) {
        this.profile = profile
        this.versionId = versionId
        this.version = profile.repository.getVersion(versionId).resolve(profile.repository)

        contentPane.children.clear()
        forge = null
        liteloader = null
        optifine = null

        for (library in version.libraries) {
            val removeAction = { _: InstallerItem ->
                val newList = LinkedList(version.libraries)
                newList.remove(library)
                VersionJSONSaveTask(profile.repository, version.copy(libraries = newList))
                        .with(task { profile.repository.refreshVersions() })
                        .with(task(Scheduler.JAVAFX) { loadVersion(this.profile, this.versionId) })
                        .start()
            }
            if (library.groupId.equals("net.minecraftforge", ignoreCase = true) && library.artifactId.equals("forge", ignoreCase = true)) {
                contentPane.children += InstallerItem("Forge", library.version, removeAction)
                forge = library.version
            } else if (library.groupId.equals("com.mumfrey", ignoreCase = true) && library.artifactId.equals("liteloader", ignoreCase = true)) {
                contentPane.children += InstallerItem("LiteLoader", library.version, removeAction)
                liteloader = library.version
            } else if ((library.groupId.equals("net.optifine", ignoreCase = true) || library.groupId.equals("optifine", ignoreCase = true)) && library.artifactId.equals("optifine", ignoreCase = true)) {
                contentPane.children += InstallerItem("OptiFine", library.version, removeAction)
                optifine = library.version
            }
        }
    }

    fun onAdd() {
        // TODO: if minecraftVersion returns null.
        val gameVersion = minecraftVersion(profile.repository.getVersionJar(version)) ?: return

        Controllers.decorator.startWizard(InstallWizardProvider(profile, gameVersion, version, forge, liteloader, optifine))
    }
}